hbase-server/src/main/java/org/apache/hadoop/hbase/master/assignment/AssignmentManager.java

```diff
@@ -1126,7 +1126,13 @@ private void updateRegionSplitTransition(final ServerName serverName, final Tran
       LOG.debug("Split request from " + serverName +
           ", parent=" + parent + " splitKey=" + Bytes.toStringBinary(splitKey));
     }
-    master.getMasterProcedureExecutor().submitProcedure(createSplitProcedure(parent, splitKey));
+    if (regionStates.getRegionState(parent).isOpened() &&
```

Comment:
This is still on a best effort basis right? Theoretically duplicate requests are still possible because the region state in AM (from previous request) can be updated after a duplicate procedure is submitted thus passing this check, although that is less likely..

Reply:
Agree. Let's add a comment here to say that this is not perfect, and is only used to reduce the concerns for operators. So later developers will know that the actual fencing is in other places.

And please get the RegionState once and store it a local variable, and then test isOpened and isSplitting on the same variable? And do we need to check whether it is null?

Reply:
> And do we need to check whether it is null?

Just to understand better, while creating Split Procedure, `regionStates.getRegionState(parent)` returning `null` should not happen right because we are trying to split an existing region (must be present in AM)? Or did I miss some race condition?

Reply:
You should also not submit the request twice right? Race condition could happen everywhere. If the region has already been split and this is a delayed request then it could be null?

Reply:
If the request is delayed to the point where parent is not only just successfully split but also removed after no refCount existing on it, then yes you are right, this is possible. Good to cover null check also.

Reply:
Do we need `!regionStates.getRegionState(parent).isSplitting()` check here? When region state is OPEN it will be always true.

Reply:
This addresses a case I've seen in real life. It's meant to fix this corner case to make the logs and operation less messy.

Reply:
Ah, now github refreshes and I see all the other comments. Ok, will improve this. Just a moment...